pkg/network/usm/shared_libraries.go

```diff
@@ -135,13 +130,77 @@ func newSOWatcher(perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
 			byID:          make(map[pathIdentifier]*soRegistration),
 			byPID:         make(map[uint32]pathIdentifierSet),
 			blocklistByID: make(pathIdentifierSet),
+
+			telemetry: soRegistryTelemetry{
+				libRegistered:               metricGroup.NewMetric("registered"),
+				libAlreadyRegistered:        metricGroup.NewMetric("already_registered"),
+				libBlocked:                  metricGroup.NewMetric("blocked"),
+				libUnregistered:             metricGroup.NewMetric("unregistered"),
+				libUnregisterNoCB:           metricGroup.NewMetric("unregister_no_callback"),
+				libUnregisterErrors:         metricGroup.NewMetric("unregister_errors"),
+				libUnregisterFailedCB:       metricGroup.NewMetric("unregister_failed_cb"),
+				libUnregisterPathIDNotFound: metricGroup.NewMetric("unregister_path_id_not_found"),
+				libHits:                     metricGroup.NewMetric("hits"),
+				libMatches:                  metricGroup.NewMetric("matches"),
+			},
 		},
 	}
 }
 
+type pathIdentifierSet = map[pathIdentifier]struct{}
+
+type soRegistryTelemetry struct {
+	// a library can be :
+	//  o Registered : it's a new library
+	//  o AlreadyRegistered : we have already hooked (uprobe) this library (unique by pathID)
+	//  o HookFailed : uprobe registration failed for one library
+	//  o Blocked : previous uprobe registration failed, so we block further call
+	//  o Unregistered : a library hook is unregistered, meaning there are no more refcount to the corresponding pathID
+	//  o UnregisterNoCB : unregister event has been done but the rule doesn't have an unregister callback
+	//  o UnregisterErrors : we encounter an error during the unregistration, looks at the logs for further details
+	//  o UnregisterFailedCB : we encounter an error during the callback unregistration, looks at the logs for further details
+	//  o UnregisterPathIDNotFound : we can't find the pathID registration, it's a bug, this value should be always 0
+	libRegistered               *telemetry.Metric
+	libAlreadyRegistered        *telemetry.Metric
+	libHookFailed               *telemetry.Metric
+	libBlocked                  *telemetry.Metric
+	libUnregistered             *telemetry.Metric
+	libUnregisterNoCB           *telemetry.Metric
+	libUnregisterErrors         *telemetry.Metric
+	libUnregisterFailedCB       *telemetry.Metric
+	libUnregisterPathIDNotFound *telemetry.Metric
+
+	// numbers of library events from the kernel filter (Hits) and matching (Matches) the registered rules
+	libHits    *telemetry.Metric
+	libMatches *telemetry.Metric
+}
+
+type soRegistry struct {
+	m     sync.RWMutex
+	byID  map[pathIdentifier]*soRegistration
+	byPID map[uint32]pathIdentifierSet
+
+	// if we can't register a uprobe we don't try more than once
+	blocklistByID pathIdentifierSet
+
+	telemetry soRegistryTelemetry
+}
+
 type soRegistration struct {
 	uniqueProcessesCount atomic.Int32
 	unregisterCB         func(pathIdentifier) error
+
+	telemetry *soRegistryTelemetry
```

Comment:
how can both soRegistration and soRegistry have the same telemetry ?

Reply:
so let's document that explicitly for making it clear for others to read

Reply:
done